ADBDEV-5423 Always materialize non-SELECT CTEs

bimboterminator1 opened the pull request:
Always materialize non-SELECT CTEs

Planner inlined each reference to CTE containing modifying DML operations into
the query plan. This inlining led to several problems. Each reference to the CTE
resulted in the operation being repeated, causing duplicating data modifications
That behaviour could lead to results that are inconsistent or erroneous.

The proposed solution involves constant materialization of such CTEs using the
SharedScan node, which functions similarly to the CTEScan node in vanilla
PostgreSQL (in local SharedInputScan case). This approach prevents the
repeated execution of the modifying operations by sharing the results of the CTE
across different parts of the query plan. To implement this change, a
modification in the set_cte_pathlist function is done. Specifically, if the
subquery's command type is not SELECT, the CTE is marked for sharing.

bimboterminator1 committed:
Throw error if one has a few refs to modifying CTE

Multiple references to CTEs which contain modifying DML operations can lead
to duplicated data modifications. This commit prohibits such cases and cuts
the query planning with the informative error. The if-clause, which checks
whether current cte is referenced more than once and contains any subquery
with any non-SELECT operation, is proposed. This solution required introduction
of a special walker that covers various cases of subqueries.

bimboterminator1 commented:
> It' an error. 

Yes, It's undesirable behaviour.
But covering such cases requires more complex changes, originally it was decided to ignore those queries. I'll investigate this issue a little bit

bandetto commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -4434,6 +4444,165 @@ is_query_contain_limit_groupby(Query *parse)
+				 * up from the current level. This level is higher than level
+				 * where cte3 is defined. cte3 is defined at level 3 and cte2
+				 * is at level 2. In this case we can derive parent root,
+				 * whose cteList contains query of cte2.
```
This comment is difficult to read. Can you split it into multiple paragraphs?

The logic seems complex. Can you explain why it is done this way and how it works in the PR description?

bimboterminator1 replied:
> Can you explain why it is done this way and how it works in the PR description?

I see no reason to explain complex code logic in patch description. Comments in the code and provided test examples are enough. The main point of the patch to prevent inlining of CTEs with DML inside. The algorithm itself is a maintenance entity, and I do not want to make patch description overcomplicated

bandetto commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -4434,6 +4444,165 @@ is_query_contain_limit_groupby(Query *parse)
+
+/*
+ * Checks whether given cte carries a non-SELECT operation inside.
+ *
```
Empty comment line.

bandetto commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -4434,6 +4444,165 @@ is_query_contain_limit_groupby(Query *parse)
+					PlannerInfo *cteroot = ctx->cteroot;
+					Index		query_level = ctx->m_query_level;
+					Index		levelsup = ctx->cteroot->query_level -
+					(ctx->m_query_level - rte->ctelevelsup);
```
Maybe replace member access with variables defined earlier?
```suggestion
					Index		query_level = ctx->m_query_level;
					Index		levelsup = cteroot->query_level -
					(query_level - rte->ctelevelsup);
```

bandetto commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -4434,6 +4444,165 @@ is_query_contain_limit_groupby(Query *parse)
+					{
+						ctx->cteroot = ctx->cteroot->parent_root;
+
+						if (!ctx->cteroot)
```
Here it's a comparison with `NULL`:
https://github.com/arenadata/gpdb/blob/dcfea40c8578208fe830a441d3b03dbbd35a72cf/src/backend/optimizer/path/allpaths.c#L4459

Perhaps you should use `... == NULL` both times for consistency and to avoid confusion with booleans?

bandetto reviewed (changes requested):
Can you add answers to these questions in your PR description:
Why duplicated DML is bad?
What's the difference in behavior before the patch and after?

Maybe use the title from the ticket? It shows the purpose more clearly. Like:
`Prohibit inlining of CTE with DML operations`

RekGRpth commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -2962,6 +2963,15 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
+		 * operations.
+		 */
+		if (cte->cterefcount > 1 &&
+			cte_query_contains_dml((Node *) cte->ctequery, cteroot))
```
Why not using approach from [6X](https://github.com/arenadata/gpdb/pull/553/files#diff-4eff9e11378a8d7655a2cd15eea18c2fa93687163c15bcdd6e4dd54818fef85fR2042)? It covers more cases

bimboterminator1 replied:
Maybe. However, there  is no much difference in execution flow from 6X version in most common cases you talking about

bimboterminator1 committed:
Cover recursive cases

bimboterminator1 committed:
Simplified walker

bimboterminator1 committed:
Refactor

bandetto commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -4434,6 +4447,141 @@ is_query_contain_limit_groupby(Query *parse)
+	{
+		CommonTableExpr *cte = (CommonTableExpr *) node;
+
+		if (cte->cterefcount && !cte->cterecursive)
```
```suggestion
		if (cte->cterefcount > 0 && !cte->cterecursive)
```

bandetto commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -4434,6 +4447,141 @@ is_query_contain_limit_groupby(Query *parse)
+		 * query of cte2.
+		 */
+		if (rte->rtekind == RTE_CTE &&
+			rte->ctelevelsup < ctx->m_query_level &&
+			(ctx->m_query_level - rte->ctelevelsup) <= ctx->cteroot->query_level)
```
Is `m_query_level` only required to check whether RTE was already processed by the walker? It looks like we can get CTE root with only RTE and the original root by just doing this, without keeping any context besides the root:
```c
while (levelsup-- > 0)
	cteroot = cteroot->parent_root;
```

Can we get rid of the `check_dml_context` struct and simplify this condition? Attached patch is how it can look like.

[0001-Simplify-the-walker.PATCH](https://github.com/arenadata/gpdb/files/15374394/0001-Simplify-the-walker.PATCH)

bimboterminator1 replied:
rte->ctelevelsup is a relative quantity. It represents the number of querly levels up from the point where this RTE is refenced to the level where its CTE is defined. This rte can be at arbitraty level. Let's consider the example:
```
explain (costs off)
with cte as (
    insert into t1 select i, i * 100 from generate_series(1,5) i
    returning *
), cte1 as (select * from t1) select * from (
with cte1 as (select * from cte)
select * from (with cte2 as (select * from cte1) select * from cte2 a join cte2 b using (i)) foo) bar;
```
cte_query_contains_dml is called on cte2's query first. ctx->cteroot->query_level=3. When walker comes across the rte corresponding to cte1, we have: rte->ctelevelsup=2. And if we go up from cteroot on 2 levels we will get to the level  1. It's the level of cte with INSERT. And there we will find the cte1 which is not even referenced anywhere, and will not detect the insert operation. It means that we need to find strict level of cte1. And without introduction of `m_query_level` it is impossible.

Second example dedicated to the problem that  [root structure is defined](https://github.com/arenadata/gpdb/blob/adb-7.1.0/src/backend/optimizer/path/allpaths.c#L2866-L2872) only for the query level where the original cte is referenced and higher. In case of nested CTEs we do not know any subroot.
```
explain (costs off)
with cte as (
    insert into t1 select i, i * 100 from generate_series(1,5) i
    returning *
),
cte2 as (
with cte3 as (select * from cte)
select (select i from cte3) from t1
)
select * from  cte2 a join cte2 b using (i);
```
In this case rte->ctelevelsup =1 for cte3, but `ctx->cteroot->query_level=1` as well. And it will require more complex logic to validate.

>Can we get rid of the check_dml_context struct and simplify this condition? 

I doubt.

bandetto commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -2962,6 +2963,18 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
+		 * multi reference and recursive cases. Otherwise, it'll cause
+		 * duplicated DML operations.
+		 */
+		if ((cte->cterefcount > 1 || cte->cterecursive) &&
```
I think recursive CTEs with RETURNING are not possible, since CTE that is RECURSIVE must be at the top level:
```sql
CREATE TABLE ins (i int);

WITH RECURSIVE cte AS (
  WITH ins AS (
    INSERT INTO ins SELECT 1 FROM INS RETURNING *
  )
  SELECT 2 UNION ALL SELECT * FROM ins
)
SELECT * FROM cte;

ERROR:  WITH clause containing a data-modifying statement must be at the top level
```

And the only method I found to put a DML statement into a top-level recursive CTE is this semi-invalid query, which causes error for this very case:
```sql
WITH RECURSIVE cte AS (
  INSERT INTO ins SELECT 1 FROM cte i RETURNING *
) SELECT 2 FROM cte;

ERROR:  recursive query "cte" must not contain data-modifying statements
```

bimboterminator1 replied:
Look at the tests section

bimboterminator1 committed:
refcount

bandetto commented on src/test/regress/expected/with_clause.out:
```diff
@@ -2355,3 +2361,96 @@ UNION ALL
+--start_ignore
+drop table if exists t1;
+NOTICE:  table "t1" does not exist, skipping
+--end_ignore
```
This block can be omitted from the output.

bandetto commented on src/test/regress/expected/with_clause_optimizer.out:
```diff
@@ -2364,3 +2370,96 @@ UNION ALL
+--start_ignore
+drop table if exists t1;
+NOTICE:  table "t1" does not exist, skipping
+--end_ignore
```
Ditto

bandetto commented on src/test/regress/expected/with_clause.out:
```diff
@@ -2355,3 +2361,96 @@ UNION ALL
+    insert into t1 select i, i * 100 from generate_series(1,5) i
+    returning *
+) select * from cte a join cte b on a.i=b.j;
+ERROR:  Multiple inlining of non-SELECT operations is prohibited (allpaths.c:2975)
```
Maybe replace literal numbers with "XXX" in both output files for a visual matchsubs clue?
```suggestion
ERROR:  Multiple inlining of non-SELECT operations is prohibited (allpaths.c:XXX)
```

bandetto commented on src/test/regress/sql/with_clause.sql:
```diff
@@ -1,3 +1,10 @@
+-- start_matchsubs
+--
+-- m/ERROR:  Multiple inlining of non-SELECT operations is prohibited \(allpaths\.c:\d+\)/
+-- s/\d+/XXX/g
+--
+-- end_matchsubs
```
Can this matchsubs block can be moved closer to the new test cases?

bimboterminator1 replied:
I think current variant can be treated as normal practice. So, moving it is quite unnecessary. One of the benefits is that other tests can extend these matchsubs without writing another block

bimboterminator1 committed:
Format tests

RekGRpth commented:
> This commit proposes the restriction, which checks checks whether

remove one of double words "checks"

RekGRpth commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -4434,6 +4447,141 @@ is_query_contain_limit_groupby(Query *parse)
+ * Returns true if any subquery of given Query has any non-SELECT commandType.
+ */
+static bool
+query_contains_dml_walker(Node *node, check_dml_context * ctx)
```
```suggestion
query_contains_dml_walker(Node *node, check_dml_context *ctx)
```

RekGRpth commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -4434,6 +4447,141 @@ is_query_contain_limit_groupby(Query *parse)
+			PlannerInfo *cteroot = ctx->cteroot;
+			Index		query_level = ctx->m_query_level;
+			Index		levelsup = cteroot->query_level -
+			(query_level - rte->ctelevelsup);
```
bad indent?

RekGRpth commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -4434,6 +4447,141 @@ is_query_contain_limit_groupby(Query *parse)
+typedef struct check_dml_context
+{
+	PlannerInfo *cteroot;
+	Index		m_query_level;	/* absolute query level */
```
may be rename it to query_level?

bimboterminator1 replied:
I my mind it's bad idea, i want to have more obvious distinguishes between current query_level  and ctx->root->query_level.

bimboterminator1 committed:
formatting

bimboterminator1 committed:
Revert "formatting"

This reverts commit a7e56580f2db769327d369e644b9061ff898dbe8.

bimboterminator1 committed:
fix formatting

RekGRpth commented:
Could we use this information to simplify checks?
https://github.com/arenadata/gpdb/blob/30d2581fe0c52a24ce113cdea42347219afa2117/src/backend/parser/parse_cte.c#L280-L289

bimboterminator1 commented:
> Could we use this information to simplify checks?

No, we can't. The main point of our checkls is to find that modifying CTE the query of the CTE, which is referenced more than 2 times. Look at the diversity of the queries in tests. I don't see any simplier options.

RekGRpth commented:
> explain (costs off)
> with cte as (
>     insert into t1 select i, i * 100 from generate_series(1,5) i
>     returning *
> ),
> cte2 as (
> with cte3 as (select * from cte)
> select (select (select i from cte3) from t2) i from t2)
> select * from  cte2 a join cte2 b using (i);

how about such case without patch?
```sql
create table t1 (i int, j int);
create table t2 (i int, j int);
insert into t1 select i, i * 100 from generate_series(1,5) i;
insert into t2 select i, i * 100 from generate_series(1,5) i;
explain (costs off)
with cte as MATERIALIZED (
    insert into t1 select i, i * 100 from generate_series(1,5) i returning *
), cte2 as (
    with cte3 as (
        select * from cte
    ) select (
        select (
            select i from cte3 join cte3 b using (i)
        ) from t2
    ) i from t2
) select * from cte2;
                                                   QUERY PLAN                                                   
----------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Seq Scan on t2
         InitPlan 2 (returns $2)  (slice2)
           ->  Gather Motion 3:1  (slice3; segments: 3)
                 InitPlan 1 (returns $1)  (slice4)
                   ->  Gather Motion 3:1  (slice5; segments: 3)
                         ->  Hash Join
                               Hash Cond: (cte3.i = b.i)
                               ->  Redistribute Motion 3:3  (slice6; segments: 3)
                                     Hash Key: cte3.i
                                     ->  Subquery Scan on cte3
                                           ->  Shared Scan (share slice:id 6:0)
                               ->  Hash
                                     ->  Redistribute Motion 3:3  (slice7; segments: 3)
                                           Hash Key: b.i
                                           ->  Subquery Scan on b
                                                 ->  Shared Scan (share slice:id 7:0)
                                                       ->  Insert on t1
                                                             ->  Redistribute Motion 1:3  (slice8; segments: 1)
                                                                   Hash Key: i.i
                                                                   ->  Function Scan on generate_series i
                 ->  Seq Scan on t2 t2_1
 Optimizer: Postgres-based planner
(23 rows)
```

bimboterminator1 commented:
> how about such case without patch?

Explain this question. I see no inlining here

RekGRpth commented:
> > how about such case without patch?
> 
> Explain this question. I see no inlining here

yes, but patch denied this query

bimboterminator1 committed:
Materialized ctes

RekGRpth commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -2962,6 +2963,18 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
+		 */
+		if ((cte->cterefcount > 1 || cte->cterecursive) &&
+			(((Query *) cte->ctequery)->commandType != CMD_SELECT ||
+			cte_query_contains_dml((Node *) cte->ctequery, cteroot)))
```
some simplification
```suggestion
			(subquery->commandType != CMD_SELECT ||
			cte_query_contains_dml((Node *) subquery, cteroot)))
```

bimboterminator1 committed:
Merge branch 'adb-7.1.0' into ADBDEV-5423

bimboterminator1 committed:
simplification

bimboterminator1 committed:
Merge branch 'adb-7.1.0' into ADBDEV-5423

bimboterminator1 committed:
fix tests

RekGRpth commented:
How about such
```diff
diff --git a/src/backend/parser/parse_relation.c b/src/backend/parser/parse_relation.c
index 7ce350881d8..9cd7f87d7ae 100644
--- a/src/backend/parser/parse_relation.c
+++ b/src/backend/parser/parse_relation.c
@@ -2206,6 +2206,11 @@ addRangeTableEntryForCTE(ParseState *pstate,
 					 parser_errposition(pstate, rv->location)));
 	}
 
+	if (cte->cterefcount > 1 || cte->cterecursive)
+		for (ParseState *parent = pstate; parent; parent = parent->parentParseState)
+			if (parent->p_hasModifyingCTE)
+				elog(ERROR, "Multiple inlining of non-SELECT operations is prohibited");
+
 	rte->coltypes = cte->ctecoltypes;
 	rte->coltypmods = cte->ctecoltypmods;
 	rte->colcollations = cte->ctecolcollations;
```
? It covers all tests, excluding materialized case

bimboterminator1 commented:
> ? It covers all tests, excluding materialized case

Relying on `p_hasModifyingCTE` case is not enough, yep. And besides materialized case there is a case when modifying CTE is not referenced at all:
```
explain (costs off)
with cte as (
    insert into t1 select i, i * 100 from generate_series(1,5) i
    returning *
), cte2 as (select * from t1)
select * from cte2 a join cte2 b on a.i=b.j;

ERROR:  Multiple inlining of non-SELECT operations is prohibited (parse_relation.c:2211)
```

It means we anyway need to check each modifying CTE for materialized and referenced features. We'll need kind of a walker. But maving a check to the parser stage is a good idea, I'll investigate this approach.

bimboterminator1 committed:
Add materialized condition

bimboterminator1 committed:
Empty line

bimboterminator1 commented:
> How about such

I decided to leave the walker at planner stage. Doing it at parsing time is less intuitive  and more complex.

Moreover, I've added condition `ignoreMaterialized` to use `cte_contain_dml` for the #964 (The best variant will be moving all the CTAS checks to the allpath.c set_cte_pathlist).

bimboterminator1 committed:
Merge branch 'adb-7.1.0' into ADBDEV-5423

bimboterminator1 committed:
Revert original solution

bimboterminator1 committed:
Materialize non-SELECT CTEs constantly

Planner inlined each reference to CTE containing modifying DML operations into
the query plan. This inlining led to several problems. Each reference to the CTE
resulted in the operation being repeated, causing duplicating data modifications
That behaviour could lead to results that are inconsistent or erroneous.

The proposed solution involves constant materialization of such CTEs using the
SharedScan node, which functions similarly to the CTEScan node in vanilla
PostgreSQL (in local SharedInputScan case). This approach prevents the
repeated execution of the modifying operations by sharing the results of the CTE
across different parts of the query plan. To implement this change, a
modification in the set_cte_pathlist function is done. Specifically, if the
subquery's command type is not SELECT, the CTE is marked for sharing.

bimboterminator1 committed:
Fix tests

RekGRpth commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -2953,6 +2953,16 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
+	 * materialize such CTEs by utilizing the existing SharedInputScan node.
+	 * This idea is similar to vanilla PostgreSQL approach.
+	 */
+	if (subquery->commandType != CMD_SELECT)
+		is_shared = true;
```
Is it possible that the code a few lines below will cancel sharing again?

bimboterminator1 replied:
No, contain_outer_selfref() checks whether cte contains **outer** self-refernce to recursive cte.  (Please, recheck)

The condition on contain_outer_selfref() prevents `z` from being shared because it contains self-refernce rte `x`.

```
explain (verbose, costs off)
with recursive x(a) as
  ((values ('a'), ('b'))
   union all
   (with z as  materialized (select * from x)
    select z.a || z1.a as a from z cross join z as z1
    where length(z.a || z1.a) < 5))
select * from x;
                    QUERY PLAN                     
---------------------------------------------------
 Recursive Union
   ->  Values Scan on "*VALUES*"
         Output: "*VALUES*".column1
   ->  Nested Loop
         Output: (x.a || x_1.a)
         Join Filter: (length((x.a || x_1.a)) < 5)
         ->  WorkTable Scan on x
               Output: x.a
         ->  WorkTable Scan on x x_1
               Output: x_1.a
 Optimizer: Postgres-based planner
 Settings: optimizer = 'off'
(12 rows)
```
In case of modifying DML we cannot achive the same structure:


```

explain (costs off)
with recursive cte as (
    select 1 as i from t1 where t1.j = 2
    union all
    select foo.j + 1 from (with cte2 as (update t1 set i = cte.i from cte where cte.i=t1.j) select * from cte2) foo where foo.i=1)
 select * from cte;
 ERROR:  WITH clause containing a data-modifying statement must be at the top level
LINE 5:     select foo.j + 1 from (with cte2 as (update t1 set i = c...
``` 

This example is also valid (cte3 does not contain outer self refences):
```
explain (costs off)
with recursive cte as (
    select 1 as i from t1 where t1.j = 2
    union all
    select foo.j + 1 from (with cte2 as (select * from cte3 join cte using (i) ) select * from cte2) foo where foo.i=1),
cte3 as (insert into t1 select i, i * 100 from generate_series(1,5) i returning *)
 select * from cte;

                                                 QUERY PLAN                                                 
------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Recursive Union
         ->  Seq Scan on t1
               Filter: (j = 2)
         ->  Subquery Scan on foo
               ->  Subquery Scan on cte2
                     ->  Nested Loop
                           ->  WorkTable Scan on cte
                                 Filter: (i = 1)
                           ->  Materialize
                                 ->  Broadcast Motion 3:3  (slice2; segments: 3)
                                       ->  Subquery Scan on cte3
                                             Filter: (cte3.i = 1)
                                             ->  Shared Scan (share slice:id 2:0)
                                                   ->  Insert on t1 t1_1
                                                         ->  Redistribute Motion 1:3  (slice3; segments: 1)
                                                               Hash Key: i.i
                                                               ->  Function Scan on generate_series i
 Optimizer: Postgres-based planner
(19 rows)

```

RekGRpth replied:
> No

Then why do we need to call the contain_outer_selfref function again if we know for sure that it will return false?

bimboterminator1 replied:
what do you suggest, move the code after this condition?

RekGRpth replied:
may be add else? like
```suggestion
	if (subquery->commandType != CMD_SELECT)
		is_shared = true;
	else if (is_shared && contain_outer_selfref(cte->ctequery))
		is_shared = false;
```

bimboterminator1 committed:
Add else

Commit:
Merge branch 'adb-7.1.0' into ADBDEV-5423

andr-sokolov commented on src/test/regress/expected/with_clause.out:
```diff
@@ -2355,3 +2355,518 @@ UNION ALL
+    38
+(1 row)
+
+-- Test multi refernce cases
```
typo "refernce"

andr-sokolov commented on src/test/regress/output/pgstat_qd_tabstat.source:
```diff
@@ -42,19 +42,20 @@ select n_tup_ins, n_tup_upd, n_tup_del, n_tup_hot_upd, n_live_tup, n_dead_tup, n
    ->  Seq Scan on table_for_initplan
- Optimizer: Postgres query optimizer
-(10 rows)
+ Optimizer: Postgres-based planner
```
Why is the optimizer name changed?

bimboterminator1 committed:
tests
